### PR TITLE
✅ Add a unit testing system

### PR DESCRIPTION
**Pull request by BattleCh1cken**
#### Summary

This pull request adds the ability to unit test parts of LemLib with the Googletest testing framework.

#### Motivation

This will help unsure the correctness of LemLib code without having to actually run it on a robot. This will not only speed up development, it will also directly deliver dopamine to my brain when I see all the tests pass.

#### Test Plan

Run the tests.

#### Additional Notes

Heavily inspired from [RaidZeroLib](https://github.com/Ryan4253/RaidZeroLib).
Thanks to @meisZWFLZ for getting started on this.

The workflow will upload coverage data to [codecov.io](https://about.codecov.io/). An access token will have to be added to the repository so the data can be properly uploaded.

I pinky promise not to create another one of these. This is the last time I swear.

## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/6951367440)
- via manual download: [LemLib@0.5.0-rc1-c0e6cd1.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1065761252.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc1-c0e6cd1.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1065761252.zip;
pros c fetch LemLib@0.5.0-rc1-c0e6cd1.zip;
pros c apply LemLib@0.5.0-rc1-c0e6cd1;
rm LemLib@0.5.0-rc1-c0e6cd1.zip;
```<!-- DO NOT REMOVE!! -->
<!-- bot: nightly-link -->
<!-- commit-sha: e65784a392bac9f48549fd76cdd6a850507870b1 -->
## Download the template for this pull request: 

> [!NOTE]  
> This is auto generated from [`Add Template to Pull Request`](https://github.com/LemLib/LemLib/actions/runs/6970742428)
- via manual download: [LemLib@0.5.0-rc1-c18f46c.zip](https://nightly.link/LemLib/LemLib/actions/artifacts/1069508238.zip)
- via PROS Integrated Terminal: 
 ```
curl -o LemLib@0.5.0-rc1-c18f46c.zip https://nightly.link/LemLib/LemLib/actions/artifacts/1069508238.zip;
pros c fetch LemLib@0.5.0-rc1-c18f46c.zip;
pros c apply LemLib@0.5.0-rc1-c18f46c;
rm LemLib@0.5.0-rc1-c18f46c.zip;
```

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -162,7 +163,21 @@ You can improve the documentation by:
 Changes should be requested via a [Pull Request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request) to the LemLib repository. You can use the [Pull Request Template](.github/PULL_REQUEST_TEMPLATE.md) to structure your pull request. Additionally, you should verify that the documentation builds correctly on your machine. You can do so by running `doxygen` in the root directory of the project. The documentation will be generated in the `docs` folder. You can then open the `index.html` file in your browser to view the documentation.
 
 
+### Writing Unit Tests
 
+Unit tests are an essential part of ensuring that your code is correct. LemLib supports unit testing with the [GoogleTest](https://google.github.io/googletest/) testing framework. The existing tests can be found in the `test` folder.
+
+<!--TODO: write file structure guidelines-->
```
Comment (liam-teale) [inline .github/CONTRIBUTING.md:170]: did you write the file structure guidelines?

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# Copied mostly from Ryan4253/RaidZeroLib: https://github.com/Ryan4253/RaidZeroLib/blob/06cc9796c5e7a1d8651d2b725cfde4c0e34d329f/CMakeLists.txt
```
Comment (liam-teale) [inline CMakeLists.txt:1]: is this compliant with the license of RaidZeroLib?

**Review by liam-teale — CHANGES_REQUESTED** [review]
Pay special attention to the license of raidzerolib

**Comment by BattleCh1cken** [comment]
IP infringement should now be resolved.

**Review by liam-teale — APPROVED** [review]
LGTM

**Pull request merged**